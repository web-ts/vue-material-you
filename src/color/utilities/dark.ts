import { DarkModeOptions, DarkModeState } from "../types";

export const darkModeOptions: DarkModeOptions = reactive({
  userMode: "auto",
  systemMode: "dark"
});

/**
 * The state of the dark mode
 */
export const dark = computed(() => {
  if (darkModeOptions.userMode === "auto") return darkModeOptions.systemMode === "dark";

  return darkModeOptions.userMode === "dark";
});

export function setDark(mode: DarkModeState) {
  darkModeOptions.userMode = mode;
}
