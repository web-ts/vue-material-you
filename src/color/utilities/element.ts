/**
 * Create the style element to be used with the color module
 */
export function createColorElement(): void {
  const target = document.getElementById("vmu-colors") as HTMLStyleElement | null;

  if (!target) {
    const style = document.createElement("style");

    style.id = "vmu-colors";
    document.head.appendChild(style);
  }
}

export function updateColorElementData(data: string) {
  const target = document.getElementById("vmu-colors") as HTMLStyleElement | null;

  if (!target) throw new Error("Could not find color style element.");

  target.innerHTML = `* {
${data}
}`;
}
