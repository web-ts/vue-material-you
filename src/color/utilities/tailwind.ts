function generateColor(colorName: string) {
  return ({ opacityVariable, opacityValue }: any) => {
    if (opacityValue !== undefined) {
      return `rgba(var(--vmu-color-${colorName}), ${opacityValue})`;
    }
    if (opacityVariable !== undefined) {
      return `rgba(var(--vmu-color-${colorName}), var(${opacityVariable}, 1))`;
    }

    return `rgb(var(--vmu-color-${colorName}))`;
  };
}

export const tailwindColors = {
  primary: generateColor("primary"),
  "on-primary": generateColor("on-primary"),
  "primary-container": generateColor("primary-container"),
  "on-primary-container": generateColor("on-primary-container"),
  secondary: generateColor("secondary"),
  "on-secondary": generateColor("on-secondary"),
  "secondary-container": generateColor("secondary-container"),
  "on-secondary-container": generateColor("on-secondary-container"),
  tertiary: generateColor("tertiary"),
  "on-tertiary": generateColor("on-tertiary"),
  "tertiary-container": generateColor("tertiary-container"),
  "on-tertiary-container": generateColor("on-tertiary-container"),
  error: generateColor("error"),
  "on-error": generateColor("on-error"),
  "error-container": generateColor("error-container"),
  "on-error-container": generateColor("on-error-container"),
  warning: generateColor("warning"),
  "on-warning": generateColor("on-warning"),
  "warning-container": generateColor("warning-container"),
  "on-warning-container": generateColor("on-warning-container"),
  success: generateColor("success"),
  "on-success": generateColor("on-success"),
  "success-container": generateColor("success-container"),
  "on-success-container": generateColor("on-success-container"),
  info: generateColor("info"),
  "on-info": generateColor("on-info"),
  "info-container": generateColor("info-container"),
  "on-info-container": generateColor("on-info-container"),
  background: generateColor("background"),
  "on-background": generateColor("on-background"),
  surface: generateColor("surface"),
  "on-surface": generateColor("on-surface"),
  "surface-variant": generateColor("surface-variant"),
  "on-surface-variant": generateColor("on-surface-variant"),
  outline: generateColor("outline"),
  shadow: generateColor("shadow"),
  "inverse-surface": generateColor("inverse-surface"),
  "inverse-on-surface": generateColor("inverse-on-surface"),
  "inverse-primary": generateColor("inverse-primary"),
  disabled: generateColor("disabled")
};
