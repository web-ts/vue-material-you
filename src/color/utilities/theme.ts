import { argbFromHex, hexFromArgb, themeFromSourceColor } from "../material-color";
import { rgbFromHex } from "./hex";

export function createTheme(color: string, dark: boolean) {
  // Create a theme.
  const theme = themeFromSourceColor(argbFromHex(color), [
    {
      name: "success",
      value: 4278229316,
      blend: false
    },
    {
      name: "info",
      value: 4284727519,
      blend: false
    },
    {
      name: "warning",
      value: 4292389403,
      blend: false
    }
  ]);

  const scheme = dark ? theme.schemes.dark : theme.schemes.light;

  const properties = new Map<string, string>();

  for (const [key, value] of Object.entries(scheme.toJSON())) {
    const token = key.replace(/([a-z])([A-Z])/g, "$1-$2").toLowerCase();
    const color = rgbFromHex(hexFromArgb(value));

    properties.set(`--vmu-color-${token}`, color);
  }

  theme.customColors.forEach((customColor) => {
    const group = dark ? customColor.dark : customColor.light;

    for (const [key, value] of Object.entries(group)) {
      const token = key
        .replace(/([a-z])([A-Z])/g, "$1-$2")
        .toLowerCase()
        .replace("color", customColor.color.name);
      const color = rgbFromHex(hexFromArgb(value));

      properties.set(`--vmu-color-${token}`, color);
    }
  });

  properties.set("--vmu-color-disabled", dark ? "227, 227, 227" : "31, 31, 31");

  return Array.from(properties)
    .map(([key, val]) => `${key}:${val};`)
    .join("\n");
}
