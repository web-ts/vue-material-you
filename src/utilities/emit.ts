/**
 * Wrapper around vue's emit declaration with support for types
 */
export default function <T>() {
  return (_v: T) => true;
}
