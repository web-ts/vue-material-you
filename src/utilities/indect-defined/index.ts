import type { InjectionKey } from "vue";

/**
 * Wrapper around vue's inject function that ensures that the injected value is defined.
 * @param key The key of the value to inject.
 */
export default function <T>(key: InjectionKey<T>) {
  const value = inject(key);

  if (value === undefined) throw new Error(`Injection key ${key.toString()} is not defined`);

  return value;
}
