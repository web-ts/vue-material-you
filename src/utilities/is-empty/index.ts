export default function (value: any) {
  if (value === null || value === undefined) return true;
  if (typeof value === "string" && value === "") return true;
  if (Array.isArray(value) && value.length === 0) return true;
  if (typeof value === "object" && Object.keys(value).length === 0) return true;

  return false;
}
