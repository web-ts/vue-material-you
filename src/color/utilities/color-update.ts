import { isHexColor } from "./hex";

import { createTheme } from "./theme";

export const colorData = reactive({
  light: "",
  dark: ""
});

export const internalColor = ref("");

export function updateColor(color: string) {
  if (!isHexColor(color)) throw new Error("Invalid HEX color. Please use a valid HEX (eg. #12F349) color.");

  const light = createTheme(color, false);
  const dark = createTheme(color, true);

  colorData.light = light;
  colorData.dark = dark;

  internalColor.value = color;
}

export const color = computed({
  get: () => internalColor.value,
  set: (newVal) => updateColor(newVal)
});
