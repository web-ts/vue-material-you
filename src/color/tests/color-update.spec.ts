import { describe, expect, it } from "vitest";

import { colorData, internalColor, updateColor, color } from "../utilities/color-update";

describe("color-update", () => {
  it("should update the color", () => {
    updateColor("#6750A4");

    expect(colorData.light).toMatchSnapshot();
    expect(colorData.dark).toMatchSnapshot();
    expect(internalColor.value).toEqual("#6750A4");
  });

  it("should throw an error for an invalid color", () => {
    try {
      updateColor("invalid");
    } catch (error: any) {
      expect(error.message).toEqual("Invalid HEX color. Please use a valid HEX (eg. #12F349) color.");
    }
  });

  it("should update the color via the computed variable", () => {
    color.value = "#6750A4";

    expect(internalColor.value).toEqual("#6750A4");
  });
});
