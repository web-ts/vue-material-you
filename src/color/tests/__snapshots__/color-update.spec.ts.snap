// Vitest Snapshot v1

exports[`color-update > should update the color 1`] = `
"--vmu-color-primary:103, 80, 164;
--vmu-color-on-primary:255, 255, 255;
--vmu-color-primary-container:233, 221, 255;
--vmu-color-on-primary-container:34, 0, 93;
--vmu-color-secondary:98, 91, 113;
--vmu-color-on-secondary:255, 255, 255;
--vmu-color-secondary-container:232, 222, 248;
--vmu-color-on-secondary-container:30, 25, 43;
--vmu-color-tertiary:126, 82, 96;
--vmu-color-on-tertiary:255, 255, 255;
--vmu-color-tertiary-container:255, 217, 227;
--vmu-color-on-tertiary-container:49, 16, 29;
--vmu-color-error:186, 26, 26;
--vmu-color-on-error:255, 255, 255;
--vmu-color-error-container:255, 218, 214;
--vmu-color-on-error-container:65, 0, 2;
--vmu-color-background:255, 251, 255;
--vmu-color-on-background:28, 27, 30;
--vmu-color-surface:255, 251, 255;
--vmu-color-on-surface:28, 27, 30;
--vmu-color-surface-variant:231, 224, 235;
--vmu-color-on-surface-variant:73, 69, 78;
--vmu-color-outline:122, 117, 127;
--vmu-color-outline-variant:202, 196, 207;
--vmu-color-shadow:0, 0, 0;
--vmu-color-scrim:0, 0, 0;
--vmu-color-inverse-surface:49, 48, 51;
--vmu-color-inverse-on-surface:244, 239, 244;
--vmu-color-inverse-primary:207, 188, 255;
--vmu-color-success:0, 110, 47;
--vmu-color-on-success:255, 255, 255;
--vmu-color-success-container:129, 252, 153;
--vmu-color-on-success-container:0, 33, 9;
--vmu-color-info:0, 103, 127;
--vmu-color-on-info:255, 255, 255;
--vmu-color-info-container:182, 234, 255;
--vmu-color-on-info-container:0, 31, 40;
--vmu-color-warning:118, 91, 0;
--vmu-color-on-warning:255, 255, 255;
--vmu-color-warning-container:255, 223, 148;
--vmu-color-on-warning-container:36, 26, 0;
--vmu-color-disabled:31, 31, 31;"
`;

exports[`color-update > should update the color 2`] = `
"--vmu-color-primary:207, 188, 255;
--vmu-color-on-primary:56, 30, 114;
--vmu-color-primary-container:79, 55, 138;
--vmu-color-on-primary-container:233, 221, 255;
--vmu-color-secondary:203, 194, 219;
--vmu-color-on-secondary:51, 45, 65;
--vmu-color-secondary-container:74, 68, 88;
--vmu-color-on-secondary-container:232, 222, 248;
--vmu-color-tertiary:239, 184, 200;
--vmu-color-on-tertiary:74, 37, 50;
--vmu-color-tertiary-container:99, 59, 72;
--vmu-color-on-tertiary-container:255, 217, 227;
--vmu-color-error:255, 180, 171;
--vmu-color-on-error:105, 0, 5;
--vmu-color-error-container:147, 0, 10;
--vmu-color-on-error-container:255, 180, 171;
--vmu-color-background:28, 27, 30;
--vmu-color-on-background:230, 225, 230;
--vmu-color-surface:28, 27, 30;
--vmu-color-on-surface:230, 225, 230;
--vmu-color-surface-variant:73, 69, 78;
--vmu-color-on-surface-variant:202, 196, 207;
--vmu-color-outline:148, 143, 153;
--vmu-color-outline-variant:73, 69, 78;
--vmu-color-shadow:0, 0, 0;
--vmu-color-scrim:0, 0, 0;
--vmu-color-inverse-surface:230, 225, 230;
--vmu-color-inverse-on-surface:49, 48, 51;
--vmu-color-inverse-primary:103, 80, 164;
--vmu-color-success:100, 222, 128;
--vmu-color-on-success:0, 57, 21;
--vmu-color-success-container:0, 83, 33;
--vmu-color-on-success-container:129, 252, 153;
--vmu-color-info:92, 213, 251;
--vmu-color-on-info:0, 53, 67;
--vmu-color-info-container:0, 78, 96;
--vmu-color-on-info-container:182, 234, 255;
--vmu-color-warning:241, 192, 52;
--vmu-color-on-warning:62, 46, 0;
--vmu-color-warning-container:89, 68, 0;
--vmu-color-on-warning-container:255, 223, 148;
--vmu-color-disabled:227, 227, 227;"
`;
